### PR TITLE
feat(web): Add stacking option to deduplication utilities

### PR DESCRIPTION
**Pull request by i-am-a-teapot**
Since stacked photos exist in immich and de-duplication only allows for deletion, I figured it would be nice to have the option to stack the pictures instead of deleting them.

I created a basic version of this functionality in this pull request. 
It (partially) addresses #2479 

Interface:
Mon Jul 15 06:07:00 PM CEST 2024

Let me know if I need to modify something, i did not fix the internationalization error in the tests, otherwise everything passed I think.

### CODE REVIEW

**Comment by alextran1502** [comment]
Thank you for working on this! I will take a look shortly

File: web/src/lib/components/utilities-page/duplicates/duplicates-compare-control.svelte [hunk 1]
```diff
@@ -7,13 +7,13 @@
...
   import { type AssetResponseDto } from '@immich/sdk';
-  import { mdiCheck, mdiTrashCanOutline } from '@mdi/js';
+  import { mdiCheck, mdiTrashCanOutline, mdiImageMultipleOutline } from '@mdi/js';
   import { onDestroy, onMount } from 'svelte';
   import { t } from 'svelte-i18n';
 
   export let assets: AssetResponseDto[];
   export let onResolve: (duplicateAssetIds: string[], trashIds: string[]) => void;
-
+  export let onStack: (stackAssets: AssetResponseDto[]) => void;
```
Comment (alextran1502) [inline web/src/lib/components/utilities-page/duplicates/duplicates-compare-control.svelte:16]:
```suggestion
  export let onStack: (assets: AssetResponseDto[]) => void;
```

File: web/src/routes/(user)/utilities/duplicates/[[photos=photos]]/[[assetId=id]]/+page.svelte [hunk 2]
```diff
@@ -63,6 +65,13 @@
     );
   };
 
+  const handleStack = async (duplicateId: string, assets: AssetResponseDto[]) => {
+    const ids = assets.flatMap((x) => x.id);
```
Comment (mertalev) [inline web/src/routes/(user)/utilities/duplicates/[[photos=photos]]/[[assetId=id]]/+page.svelte:69]: I think `map` should be fine here.

File: web/src/routes/(user)/utilities/duplicates/[[photos=photos]]/[[assetId=id]]/+page.svelte [hunk 3]
```diff
@@ -116,6 +125,29 @@
       $t('confirm'),
     );
   };
+
+  const handleStackAll = async () => {
+    const ids = data.duplicates.flatMap((group) => group.assets.map((asset) => asset.id));
+    let prompt, confirmText;
+    if ($featureFlags.trash) {
```
Comment (mertalev) [inline web/src/routes/(user)/utilities/duplicates/[[photos=photos]]/[[assetId=id]]/+page.svelte:132]: This doesn't seem right. Stacking has nothing to do with trashing.

File: web/src/routes/(user)/utilities/duplicates/[[photos=photos]]/[[assetId=id]]/+page.svelte [hunk 4]
```diff
@@ -132,6 +164,12 @@
         {$t('keep_all')}
       </div>
     </LinkButton>
+    <LinkButton on:click={() => handleStackAll()} disabled={!hasDuplicates}>
```
Comment (mertalev) [inline web/src/routes/(user)/utilities/duplicates/[[photos=photos]]/[[assetId=id]]/+page.svelte:167]: I think it should be disabled if more than one asset is selected and should make the selected asset the primary asset of the stack.

**Review by mertalev — COMMENTED** [review]
Thanks for working on this! I'm sure a lot of people will love this feature.

How do you feel about the "select keep all" / "select trash all" buttons? I think they feel weird now since keeping and trashing aren't the only actions. IMO it'd be best to just fold them into one button that says "select all" if not all assets are selected, and "unselect all" if all of them are.

File: web/src/routes/(user)/utilities/duplicates/[[photos=photos]]/[[assetId=id]]/+page.svelte [hunk 5]
```diff
@@ -116,6 +125,29 @@
...
+
+    return withConfirmation(
+      async () => {
+        for (let i in data.duplicates) {
+          //let parentId = suggestDuplicateByFileSize(data.duplicates[i].assets).id;
+          await stackAssets(data.duplicates[i].assets);
+          await updateAssets({
+            assetBulkUpdateDto: { ids: data.duplicates[i].assets.map((asset) => asset.id), duplicateId: null },
+          });
+        }
```
[the remark this answers is not in the document]
Comment (i-am-a-teapot) [reply web/src/routes/(user)/utilities/duplicates/[[photos=photos]]/[[assetId=id]]/+page.svelte:145]: I totally agree, but I don't think the server side supports this and I did not feel too confident touching that yet - but I will look into it and try to do this!

**Comment by i-am-a-teapot** [comment]
I created some new endpoints in the server to handle the one request stackAll operation. I hope I captured all the functionality of the old updateAll method. I am not certain if the two DTO's are necessary but it seems to work for now.
If you agree on the general structure, I think writing some tests would be necessary to ensure it works as it should.

I did not look into the app development yet, but the stack functionality will also need to change there.

**Comment by alextran1502** [comment]
Sorry for the back-and-forth on this PR. We are refactoring the stack mechanism, so I would avoid changing the code on the server for stacking. Would it be possible to use the current stack endpoint and mechanism for this purpose?

**Comment by i-am-a-teapot** [comment]
It would be possible, but we would need to live with multiple requests for stacking all duplicates.
(Pretty much like in the first commit)

**Comment by mertalev** [comment]
That would be unfortunate. If we did it one-by-one, the user has 50k duplicate groups, and handling each group takes 20ms, it would take over 16 minutes to get through all of them. The user has to wait on that page that long and not assume that clicking the button was enough, or that something went wrong and they need to refresh. I'd almost rather not have the Stack All button in that case.

**Comment by alextran1502** [comment]
Yeah, no stack all button is fine

**Comment by i-am-a-teapot** [comment]
I updated it to the best of my knowledge, it is a bit sad that the stack all button is not possible yet, but I understand.
Let me know if you still need changes.

File: web/src/lib/utils/asset-utils.ts [hunk 6]
```diff
@@ -324,13 +324,17 @@ export const getSelectedAssets = (assets: Set<AssetResponseDto>, user: UserRespo
   return ids;
 };
 
-export const stackAssets = async (assets: AssetResponseDto[]) => {
+export const stackAssets = async (assets: AssetResponseDto[], parent?: AssetResponseDto, showGoToButton?: boolean) => {
```
Comment (alextran1502) [inline web/src/lib/utils/asset-utils.ts:327]: Why do we need to change the logic of this method?

File: web/src/lib/utils/asset-utils.ts [hunk 7]
```diff
@@ -361,17 +365,19 @@ export const stackAssets = async (assets: AssetResponseDto[]) => {
   parent.stack ??= [];
   parent.stack = parent.stack.concat(children, grandChildren);
   parent.stackCount = parent.stack.length + 1;
-
-  notificationController.show({
+  const notificationObject = {
```
Comment (alextran1502) [inline web/src/lib/utils/asset-utils.ts:368]: What are we trying to achieve here? The `button` property is also not a public property AFAIK

File: web/src/routes/(user)/utilities/duplicates/[[photos=photos]]/[[assetId=id]]/+page.svelte [hunk 8]
```diff
@@ -141,6 +162,36 @@
       $t('confirm'),
     );
   };
+
```
Comment (alextran1502) [inline web/src/routes/(user)/utilities/duplicates/[[photos=photos]]/[[assetId=id]]/+page.svelte:165]: I don't think we need to keep this commented code here

File: web/src/routes/(user)/utilities/duplicates/[[photos=photos]]/[[assetId=id]]/+page.svelte [hunk 9]
```diff
@@ -157,6 +208,16 @@
         {$t('keep_all')}
       </div>
     </LinkButton>
+    <!--
```
Comment (alextran1502) [inline web/src/routes/(user)/utilities/duplicates/[[photos=photos]]/[[assetId=id]]/+page.svelte:211]: I don't think we need to keep this commented code here

**Review by alextran1502 — COMMENTED** [review]
Thank you for your work, I add some comments on a few things

**Review by alextran1502 — APPROVED** [review]

**Pull request merged**